### ISSUE TITLE
weird behavior while scaling the screen

### ISSUE THREAD

**Issue by foodtooth** [faho/kwin-tiling#80]
I'm getting weird tiling while scaling the screen to 1.5.

I'm using Arch Linux, newest plasma, 'Scale Display' function.

It makes some of the windows run out of the edge of the screen, and went black after trying to drag the window.

How can I fix this?

**Comment by faho** [comment]
I'm actually using the same scale, same distro, and it works for me.

My first idea is to check what resolution KWin reports to the script. In contents/code/util.js there is the "getTilingArea" function, which uses `workspace.clientArea(KWin.PlacementArea, screen, desktop)`.

Also, are you using multiple monitors?

**Comment by foodtooth** [comment]
I'm using single monitor on iMac. The resolution is 3840x2160.

I'm not familiar with kwin, so did a little google and used `wm console` to test the code. `KWin.PlacementArea` gives me `0`. But I don't know how to get `screen` and `desktop` argument.

Is this helpful?

**Comment by faho** [comment]
>Is this helpful?

Not by itself, no. "Kwin.PlacementArea" is just a constant so that "clientArea" returns a specific _kind_ of clientArea (there's also "MaximizeArea" and such). But it tells me that you are capable of running commands, so I'd like to know what

```js
workspace.clientArea(KWin.PlacementArea, 0, 0);
```

returns (IIRC kwin numbers both screen and desktop from 0 - if there's an issue here try with 1 instead). If it just says something like "Qt.Rect" or something, I'd like the ".x", ".y", ".width" and ".height" members of that.

Also, what kind of panel setup do you use? Where is it (are they?), is it set to autohide?

**Comment by foodtooth** [comment]
Thanks a lot. Thanks for the explanation.

## Here is the code and the output
```javascript
Code:
...
util.getTilingArea = function(screen, desktop) {
    var cA = workspace.clientArea(KWin.PlacementArea, screen, desktop);
    print("clientArea: " + JSON.stringify(cA));
    return util.copyRect(cA);
};
util.getTilingArea(0, 0);
---
Output:
clientArea: {"x":0,"y":0,"width":3840,"height":2124}
```

## About panel setup
It's a fresh new system. I haven't customized the panel. It remains the default value "Always Visible". And the config file is:
```ini
$ cat .config/plasmashellrc 
[InteractiveConsole]
Geometry=\x01���\x00\x02\x00\x00\x00\x00\x05�\x00\x00\x02\x14\x00\x00\t�\x00\x00\x06\x14\x00\x00\x05�\x00\x00\x02=\x00\x00\t�\x00\x00\x06\x0e\x00\x00\x00\x00\x00\x00\x00\x00\x0f\x00
SplitterState=\x00\x00\x00�\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x06�\x00\x00\x01\x10\x01����\x01\x00\x00\x00\x02\x00

[KFileDialog Settings]
Recent Files[$e]=file:$HOME/workspace/test/kwin-tiling-test.js,file:$HOME/workspace/test/kwin-tiling-test.js,file:$HOME/.local/share/plasmashell/interactiveconsoleautosave.js
Recent URLs[$e]=file:$HOME/workspace/test/,file:$HOME/workspace/,file:$HOME/.local/share/plasmashell/,file:$HOME/Downloads/mike/

[PlasmaRunnerManager]
LaunchCounts=3 services_konqbrowser.desktop,1 places_remote:/,13 services_systemsettings.desktop,22 services_org.kde.konsole.desktop,3 services_org.kde.kleopatra.desktop,3 services_nic.desktop,6 services_chromium.desktop,7 services_kwalletconfig5.desktop,5 services_kwinscripts.desktop,1 services_org.kde.kwalletmanager5.desktop,1 services_proxy.desktop,1 services_khotkeys.desktop,1 services_fonts.desktop,1 services_kcm_networkmanagement.desktop,1 services_org.kde.dolphin.desktop,2 services_org.kde.kgpg.desktop,2 services_shadowsocks-qt5.desktop
pluginWhiteList=services,places,shell,bookmarks,recentdocuments,locations,baloosearch

[PlasmaViews][Panel 2]
panelVisibility=0

[PlasmaViews][Panel 2][Horizontal3840]
thickness=36

[ScreenConnectors]
0=eDP

[Updates]
performed=/usr/share/plasma/shells/org.kde.plasma.desktop/contents/updates/obsolete_kickoffrc.js
```

## In addition
The script works perfect in my system while using scale 1.4 and below, but goes weird scaling to 1.5 and above.

**Comment by faho** [comment]
Sorry for the delay, I've had a bit of a busy week!

>clientArea: {"x":0,"y":0,"width":3840,"height":2124}

That actually seems correct - it means you have a 36px panel at the bottom.

I'm not sure how the scaling factor impacts this - we so far don't care about it at all since it seems like it should only apply to the window _content_ - when we set a window to be 200 x 200 px, it should get that size and not 300 x 300, no matter the scaling.

It's possible that we generate a non-int for a dimension at some point, which then gets compared to produce weird results. That might look like what you are seeing.

Would it be possible for you to take a screencast or at least a screenshot, so I have a better idea of what you're seeing?

**Comment by foodtooth** [comment]
Thanks for the reply! Here is the video:

[![weird behavior while scaling the screen](https://i.vimeocdn.com/video/629711481_640.webp)](https://vimeo.com/213476832 "weird behavior while scaling the screen - Click to Watch!")

Besides, I can feel some delay of the window when I was trying to 'close window' or 'put the window down', which didn't happen if I use scale 1.4. But I'm not sure whether it's relevant to the issue.

Thanks again for your time.

**Comment by faho** [comment]
Sorry for the delay, but GOOD NEWS!

I've finally managed to reproduce this - the issue was that with a scaling factor of 1.5, the "basicUnit" as kwin calls it (ICCCM calls it "PResizeInc" - it's a "this client only wants to grow/shrink by multiples of this) is increased to 2 (from 1). This caused our special calculations to kick off, and those calculations were wrong - there was a stray "Math.abs" in there that resulted in the geometry growing instead of shrinking.

I believe this is now fixed.
